### PR TITLE
Add support for git stash indicator

### LINKED ISSUE #19: Add `git stash` indicator
This is a nice feature that the [`pure`](https://github.com/pure-fish/pure) prompt has. It would be easy to adapt their implementation ([here](https://github.com/pure-fish/pure/blob/440dcb3adcb5ab5f134a7f7079a5cea7c1c1c44c/functions/_pure_prompt_git_stash.fish)) into this prompt.

How it looks in `pure`, excuse the color scheme. Just installed it for this example screenshot
Screen Shot 2021-03-11 at 13 48 56

Not sure what the contribution process is for this repo, but happy to open a PR if this would be accepted!

### PR DESCRIPTION
**Pull request by alex-grover**
Screen Shot 2021-09-17 at 10 48 14

closes #19.

@scorphus

### CODE REVIEW

**Comment by alex-grover** [comment]
bump @scorphus, just want to make sure you saw this 🙂

**Comment by alex-grover** [comment]
Screen Shot 2022-02-22 at 13 00 24

Example of how this feature works when stash/dirty interact. I based this on how the feature worked with other prompts that I've used, but open to alternative suggestions!

**Comment by scorphus** [comment]
Yup, just as I'd expect form looking at the code. So, even when there's things stashed but the working dir is dirty, the dirty status have precedence. I would expect to see both, otherwise, instead of just the dirty status. But, if from your experience that's what is usually done, and also that's what you'd expect, I'm okay with it.

**Comment by scorphus** [comment]
And thanks for making the screenshot, I was sure it would be an easy one for you (hope it was)

**Comment by alex-grover** [comment]
haha yep i have been using this for the past few months!

I was actually incorrect, just tested with `pure` and it does indeed show both indicators:

Screen Shot 2022-02-22 at 13 08 54

Will update the PR to reflect that! Maybe I remembered the other way from some other prompt back in my zsh days

**Comment by alex-grover** [comment]
Screen Shot 2022-02-22 at 17 48 25

Updated to display both stash and dirty indicators, spent a while staring at the conditional, not sure if there is an obvious way to simplify it and include the extra space that I'm missing 😅 

Both indicators take precedence over ahead/behind/diverged. `pure` shows all the applicable indicators at once but I personally like how the default theme currently works better so I left as is.

I also snuck in a fix for an extra space after the diverged indicator that I noticed while implementing.

File: functions/fish_prompt.fish [hunk 1]
```diff
@@ -41,7 +43,11 @@ function fish_prompt
     echo -n -s " " $directory_color $cwd $normal_color
     echo -n -s " on " $repository_color (git_branch_name) $normal_color " "
 
-    if git_is_touched
+    if git_is_touched; and test "$theme_stash_indicator" = 'yes'; and git_is_stashed
+      echo -n -s $stash " " $dirty
+    else if test "$theme_stash_indicator" = 'yes'; and git_is_stashed
+      echo -n -s $stash
+    else if git_is_touched
```
Comment (scorphus) [inline functions/fish_prompt.fish]: I'd say this is readable and explicit. I don't think simplifying would make it more readable/maintainable. Some repetition is good repetition.

You have any ideas, @faho?

Comment (faho) [reply functions/fish_prompt.fish]: The issue is `git_is_touched` calls `git diff-index` and `git diff`, which are fairly expensive as they need to scan the entire repo. This potentially calls them twice.

I'd probably do it like this:

```fish
if test "$theme_stash_indicator" = yes; and git_is_stashed
  if git_is_touched
    echo -n -s $stash " " $dirty
  else
    echo -n -s $stash
  end
else if git_is_touched
  # ...
end
```

Alternatively sorting the first if-condition as `test "$theme_stash_indicator" = yes; and git_is_stashed; and git_is_touched` would at least not call that twice, but could still check `git_is_stashed` twice. It's cheaper, but better not to do that work if you don't need to.

Comment (faho) [reply functions/fish_prompt.fish]: Alternatively alternatively, you'd build up the string to `echo` and then print it in one go (without `-s` so you get a nice space between the indicators):

```fish
set -l list
if test "$theme_stash_indicator" = yes; and git_is_stashed
    set list $list $stash
end
if git_is_touched
    set list $list $dirty
end

echo -n $list
```

Comment (alex-grover) [reply functions/fish_prompt.fish]: perfect, the latter is what i was trying to do initially but i didn't think to look at what `-s` does 🤦🏽‍♂️ 

Updated the PR and here's one more screenshot to verify, it works the same as the last version:

Screen Shot 2022-02-25 at 12 14 43

**Review by faho — APPROVED** [review]
Looks okay to me, from a cursory glance

**Comment by alex-grover** [comment]
@scorphus this is ready to merge with your review, let me know if you need anything else from me!

**Review by scorphus — APPROVED** [review]
Thanks for contributing, @alex-grover!

**Pull request merged**